### PR TITLE
Add repeat while to dispatcher

### PR DESCRIPTION
**Pull request by joaomanita**
# Description

Include a summary of the changes here.

## Checklist

- [ ] Self-review changes.
- [ ] Evaluate impact on the documentation.
- [ ] Ensure test coverage.
- [ ] Write new samples.

### CODE REVIEW

File: core/src/ecs/system/dispatcher.cpp [hunk 1]
```diff
@@ -62,11 +68,35 @@ void Dispatcher::tagSetBeforeTag(const std::string& tag)
...
+
 void Dispatcher::systemAddTag(const std::string& tag)
 {
     ENSURE_CURR_SYSTEM();
     ENSURE_TAG_SETTINGS(tag);
     mCurrSystem->tags.insert(tag);
+    std::shared_ptr<GroupStep> group = mMainStep->findGroup(tag);
+    if (group != nullptr)
+    {
+        mCurrSystem->grouptag = tag;
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:86]:
```suggestion
    {
        CUBOS_ASSERT(mCurrSystem->grouptag == nullptr, "Systems can only be tagged directly with at most one repeating tag");
        mCurrSystem->grouptag = tag;
```
We should check this right?

File: core/src/ecs/system/dispatcher.cpp [hunk 2]
```diff
@@ -62,11 +68,35 @@ void Dispatcher::tagSetBeforeTag(const std::string& tag)
...
+        mCurrGroup = group;
+    }
+
+    if (tag == "last")
+    {
+        // mMainStep->printStructure();
+    }
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:93]: Remove unused testing code

File: core/src/ecs/system/dispatcher.cpp [hunk 3]
```diff
@@ -62,11 +68,35 @@ void Dispatcher::tagSetBeforeTag(const std::string& tag)
     mTagSettings[tag]->after.tag.push_back(mCurrTag);
 }
 
+void Dispatcher::groupAddGroup()
+{
+    // CUBOS_WARN("{}", mCurrGroup->getGroupTag());
+    mCurrGroup = mCurrGroup->addGroupStep(mCurrTag, mCurrGroup);
+    // CUBOS_WARN("{}", mCurrGroup->getGroupTag());
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:75]: Remove commented testing code

File: core/src/ecs/system/dispatcher.cpp [hunk 4]
```diff
@@ -62,11 +68,35 @@ void Dispatcher::tagSetBeforeTag(const std::string& tag)
...
+    if (tag == "last")
+    {
+        // mMainStep->printStructure();
+    }
+}
+
+void Dispatcher::systemAddGroup(const std::string& grouptag)
+{
+    ENSURE_CURR_SYSTEM();
+    mCurrSystem->grouptag = grouptag;
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:99]: Rename `grouptag` to `groupTag`.

File: core/src/ecs/system/dispatcher.cpp [hunk 5]
```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
         node.m = DFSNode::BLACK;
         if (node.s != nullptr)
         {
-            mSystems.push_back(node.s);
+            // mSystems.push_back(node.s);
+            // CUBOS_WARN("{}", node.s->grouptag);
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:254]: Remove testing comments

File: core/src/ecs/system/dispatcher.cpp [hunk 6]
```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
...
         return false;
     }
     }
     return false;
 }
 
-void Dispatcher::callSystems(CommandBuffer& cmds)
+void Dispatcher::SystemStep::call(CommandBuffer& cmds, std::vector<ecs::System<bool>>& mConditions,
+                                  std::bitset<CUBOS_CORE_DISPATCHER_MAX_CONDITIONS> mRunConditions,
+                                  std::bitset<CUBOS_CORE_DISPATCHER_MAX_CONDITIONS> mRetConditions)
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:265]: Remove the `m` from the arguments. The `m` prefix is used to indicate that its a field, but here the variables are arguments.

File: core/src/ecs/system/dispatcher.cpp [hunk 7]
```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
...
+        mSystem->system.run(cmds);
+    }
+
+    // TODO: Check synchronization concerns when this gets multithreaded
+    cmds.commit();
+}
 
+void Dispatcher::GroupStep::call(CommandBuffer& cmds, std::vector<ecs::System<bool>>& mConditions,
+                                 std::bitset<CUBOS_CORE_DISPATCHER_MAX_CONDITIONS> mRunConditions,
+                                 std::bitset<CUBOS_CORE_DISPATCHER_MAX_CONDITIONS> mRetConditions)
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:309]: Same thing here regarding the argument naming

File: core/src/ecs/system/dispatcher.cpp [hunk 8]
```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
...
+                {
+                    // CUBOS_WARN("{}", SystemStepPtr->getSystem()->grouptag);
+                    systemStepPtr->call(cmds, mConditions, mRunConditions, mRetConditions);
+                }
+            }
+        }
+        if (grouptag == "main")
+        {
+            canRun = false;
         }
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:363]: What is this?

File: core/src/ecs/system/dispatcher.cpp [hunk 9]
```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
...
+
+void Dispatcher::callSystems(CommandBuffer& cmds)
+{
+
+    // Clear conditions bitmasks
+    mRunConditions.reset();
+    mRetConditions.reset();
+
+    mMainStep->call(cmds, mConditions, mRunConditions, mRetConditions);
+}
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:375]: Add missing new line at the end of file.

File: core/src/ecs/system/dispatcher.cpp [hunk 10]
```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
...
         }
-
-        // TODO: Check synchronization concerns when this gets multithreaded
-        cmds.commit();
     }
 }
+
+void Dispatcher::callSystems(CommandBuffer& cmds)
+{
+
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:369]:
```suggestion
```
Remove blank line

File: core/src/ecs/system/dispatcher.cpp [hunk 11]
```diff
@@ -220,61 +250,126 @@ bool Dispatcher::dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes)
...
+                if (auto groupStepPtr = std::dynamic_pointer_cast<GroupStep>(step))
+                {
+                    // CUBOS_WARN("{}", groupStepPtr->grouptag);
+                    groupStepPtr->call(cmds, mConditions, mRunConditions, mRetConditions);
+                }
+                else if (auto systemStepPtr = std::dynamic_pointer_cast<SystemStep>(step))
+                {
+                    // CUBOS_WARN("{}", SystemStepPtr->getSystem()->grouptag);
+                    systemStepPtr->call(cmds, mConditions, mRunConditions, mRetConditions);
+                }
```
Comment (RiscadoA) [inline core/src/ecs/system/dispatcher.cpp:357]:
```suggestion
                step->call(cmds, mConditions, mRunConditions, mRetConditions);
```

File: core/tests/ecs/dispatcher.cpp [hunk 12]
```diff
@@ -39,6 +39,18 @@ static bool pushToOrderAndSucceed(std::vector<int>& order)
     return true;
 }
 
+static bool succeed2Times(int& counter)
+{
+
```
Comment (RiscadoA) [inline core/tests/ecs/dispatcher.cpp:44]:
```suggestion
```

File: core/tests/ecs/dispatcher.cpp [hunk 13]
```diff
@@ -221,4 +234,43 @@ TEST_CASE("ecs::Dispatcher")
...
+        dispatcher.addSystem(wrapSystem(pushToOrder<1>));
+        dispatcher.systemAddGroup("repeat");
+
+        singleDispatch(dispatcher, cmdBuffer);
+        assertOrder(world, {1, 1, 1});
+    }
+
+    SUBCASE("repeat while embedded")
+    {
+        // Tests if #413 has been fixed.
```
Comment (RiscadoA) [inline core/tests/ecs/dispatcher.cpp:252]:
```suggestion
```

File: core/tests/ecs/dispatcher.cpp [hunk 14]
```diff
@@ -221,4 +234,43 @@ TEST_CASE("ecs::Dispatcher")
         singleDispatch(dispatcher, cmdBuffer);
         assertOrder(world, {1, 3});
     }
+
+    SUBCASE("repeat while")
+    {
+        // Tests if #413 has been fixed.
```
Comment (RiscadoA) [inline core/tests/ecs/dispatcher.cpp:240]:
```suggestion
```

File: core/tests/ecs/dispatcher.cpp [hunk 15]
```diff
@@ -221,4 +234,43 @@ TEST_CASE("ecs::Dispatcher")
...
+        dispatcher.addTag("repeat");
+        dispatcher.tagRepeatWhile(wrapCondition(succeed2Times));
+        dispatcher.addSystem(wrapSystem(pushToOrder<1>));
+        dispatcher.systemAddGroup("repeat");
+
+        singleDispatch(dispatcher, cmdBuffer);
+        assertOrder(world, {1, 1, 1});
+    }
+
+    SUBCASE("repeat while embedded")
```
Comment (RiscadoA) [inline core/tests/ecs/dispatcher.cpp:250]:
```suggestion
    SUBCASE("repeat while inside repeat while")
```

File: engine/src/physics/plugin.cpp [hunk 16]
```diff
@@ -71,6 +71,7 @@ void cubos::engine::physicsPlugin(Cubos& cubos)
     cubos.addPlugin(gravityPlugin);
 
     // executed every frame
+
```
Comment (RiscadoA) [inline engine/src/physics/plugin.cpp:74]:
```suggestion
```

File: core/include/cubos/core/ecs/cubos.hpp [hunk 17]
```diff
@@ -80,6 +85,12 @@ namespace cubos::core::ecs
         template <typename F>
         TagBuilder& runIf(F func);
 
+        /// @brief Adds a repeat condition to the current tag, also making it a group.
```
Comment (RiscadoA) [inline core/include/cubos/core/ecs/cubos.hpp:88]:
```suggestion
        /// @brief Makes all systems within the tag repeat as a group until the given condition evaluates to false.
```

File: core/include/cubos/core/ecs/cubos.hpp [hunk 18]
```diff
@@ -72,6 +72,11 @@ namespace cubos::core::ecs
         /// @return Reference to this object, for chaining.
         TagBuilder& after(const std::string& tag);
 
+        /// @brief Makes the current tag inherit from given tag.
```
Comment (RiscadoA) [inline core/include/cubos/core/ecs/cubos.hpp:75]:
```suggestion
        /// @brief Tags all systems on this tag with the given tag.
```

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: core/src/ecs/system/dispatcher.cpp (hunk 8 above)
Comment (joaomanita) [reply core/src/ecs/system/dispatcher.cpp:363]: main is the only group that doesn't have a condition attached to it, if this wasn't there the loop  would repeat forever.

**Review by RiscadoA — APPROVED** [review]
Thanks for working on this! Will be a great help in physics. Sorry for being so picky with the reviews, but we must make sure this part of the code works properly. Feel free to merge whenever you're ready :rocket:

**Pull request merged**